<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to influential • influential</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to influential">
<meta property="og:description" content="influential">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">influential</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.3.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Vignettes.html">Introduction to influential</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/asalavaty/influential/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="Vignettes_files/header-attrs-2.11/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to influential</h1>
            <h3 data-toc-skip class="subtitle"><em>Identification and classification of the most influential nodes</em></h3>
                        <h4 data-toc-skip class="author">Adrian (Abbas) Salavaty 20-Oct-2021</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/asalavaty/influential/blob/master/vignettes/Vignettes.Rmd" class="external-link"><code>vignettes/Vignettes.Rmd</code></a></small>
      <div class="hidden name"><code>Vignettes.Rmd</code></div>

    </div>

    
    
<style>
body {
text-align: justify}
</style>
<div id="overview" class="section level1">
<h1 class="hasAnchor">
<a href="#overview" class="anchor" aria-hidden="true"></a>Overview</h1>
<p><img src="../man/figures/Symbol.png" width="25%" style="float:right; padding:10px"></p>
<p><code>influential</code> is an R package mainly for the identification of the most influential nodes in a network as well as the classification and ranking of top candidate features. The <code>influential</code> package contains several functions that could be categorized into five groups according to their purpose:</p>
<ul>
<li>Network reconstruction</li>
<li>Calculation of centrality measures</li>
<li>Assessment of the association of centrality measures</li>
<li>Identification of the most <code>influential</code> network nodes</li>
<li>Experimental data-based classification and ranking of features</li>
</ul>
<p>The sections below introduce these five categories. However, if you wish not going through all of the functions and their applications, you may skip to any of the novel methods proposed by the <code>influential</code>, including:</p>
<ul>
<li><a href="#ExIR">ExIR</a></li>
<li><a href="#IVI">IVI</a></li>
<li><a href="#HubnessScore">Hubness score</a></li>
<li><a href="#SpreadingScore">Spreading score</a></li>
<li><a href="#SIRIR">SIRIR</a></li>
<li><a href="#CompManipulate">Computational manipulation</a></li>
</ul>
<hr>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/asalavaty/influential" class="external-link">influential</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="network-reconstruction" class="section level1">
<h1 class="hasAnchor">
<a href="#network-reconstruction" class="anchor" aria-hidden="true"></a>Network reconstruction</h1>
<p>Three functions have been obtained from the <code>igraph</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> R package for the reconstruction of networks.</p>
<div id="from-a-data-frame" class="section level2">
<h2 class="hasAnchor">
<a href="#from-a-data-frame" class="anchor" aria-hidden="true"></a>From a data frame</h2>
<p>In the data frame the first and second columns should be composed of source and target nodes.<br>
A sample appropriate data frame is brought below:</p>
<table class="table">
<thead><tr class="header">
<th align="left">lncRNA</th>
<th align="left">Coexpressed.Gene</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ADAMTS9-AS2</td>
<td align="left">A2M</td>
</tr>
<tr class="even">
<td align="left">ADAMTS9-AS2</td>
<td align="left">ABCA6</td>
</tr>
<tr class="odd">
<td align="left">ADAMTS9-AS2</td>
<td align="left">ABCA8</td>
</tr>
<tr class="even">
<td align="left">ADAMTS9-AS2</td>
<td align="left">ABCA9</td>
</tr>
<tr class="odd">
<td align="left">ADAMTS9-AS2</td>
<td align="left">ABI3BP</td>
</tr>
<tr class="even">
<td align="left">ADAMTS9-AS2</td>
<td align="left">AC093110.3</td>
</tr>
</tbody>
</table>
<p>This is a co-expression dataset obtained from a paper by Salavaty <strong>et al</strong>.<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span>d<span class="op">=</span><span class="va">MyData</span><span class="op">)</span></code></pre></div>
<p>If you look at the class of <code>My_graph</code> you should see that it has an <code>igraph</code> class:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>
<span class="co">#&gt; [1] "igraph"</span></code></pre></div>
</div>
<div id="from-an-adjacency-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#from-an-adjacency-matrix" class="anchor" aria-hidden="true"></a>From an adjacency matrix</h2>
<p>A sample appropriate adjacency matrix is brought below:</p>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">LINC00891</th>
<th align="right">LINC00968</th>
<th align="right">LINC00987</th>
<th align="right">LINC01506</th>
<th align="right">MAFG-AS1</th>
<th align="right">MIR497HG</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">LINC00891</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">LINC00968</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">LINC00987</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">LINC01506</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">MAFG-AS1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">MIR497HG</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<ul>
<li>Note that the matrix has the same number of rows and columns.</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.adjacency</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_adjacency_matrix.html">graph_from_adjacency_matrix</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        </code></pre></div>
</div>
<div id="from-an-incidence-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#from-an-incidence-matrix" class="anchor" aria-hidden="true"></a>From an incidence matrix</h2>
<p>A sample appropriate incidence matrix is brought below:</p>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Gene_1</th>
<th align="right">Gene_2</th>
<th align="right">Gene_3</th>
<th align="right">Gene_4</th>
<th align="right">Gene_5</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">cell_1</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">cell_2</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">cell_3</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">cell_4</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_adjacency_matrix.html">graph_from_adjacency_matrix</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        </code></pre></div>
</div>
<div id="from-a-sif-file" class="section level2">
<h2 class="hasAnchor">
<a href="#from-a-sif-file" class="anchor" aria-hidden="true"></a>From a SIF file</h2>
<p>SIF is the common output format of the Cytoscape software.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sif2igraph.html">sif2igraph</a></span><span class="op">(</span>Path <span class="op">=</span> <span class="st">"Sample_SIF.sif"</span><span class="op">)</span>        

<span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>
<span class="co">#&gt; [1] "igraph"</span></code></pre></div>
<p><a href="#top">Back to top</a></p>
</div>
</div>
<div id="calculation-of-centrality-measures" class="section level1">
<h1 class="hasAnchor">
<a href="#calculation-of-centrality-measures" class="anchor" aria-hidden="true"></a>Calculation of centrality measures</h1>
<p>To calculate the centrality of nodes within a network several different options are available. The following sections describe how to obtain the names of network nodes and use different functions to calculate the centrality of nodes within a network. Although several centrality functions are provided, we recommend the <a href="#IVI">IVI</a> for the identification of the most <code>influential</code> nodes within a network.</p>
<blockquote>
<p>By the way, the results of all of the following centrality functions could be conveniently illustrated using the <a href="#NetVis">centrality-based network visualization function</a>.</p>
</blockquote>
<div id="network-vertices" class="section level2">
<h2 class="hasAnchor">
<a href="#network-vertices" class="anchor" aria-hidden="true"></a>Network vertices</h2>
<p>Network vertices (nodes) are required in order to calculate their centrality measures. Thus, before calculation of network centrality measures we need to obtain the name of required network vertices. To this end, we use the <code>V</code> function, which is obtained from the <code>igraph</code> package. However, you may provide a character vector of the name of your desired nodes manually.</p>
<ul>
<li>Note in many of the centrality index functions the entire network nodes are assessed if no vector of desired vertices is provided.</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">My_graph_vertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">My_graph_vertices</span><span class="op">)</span>
<span class="co">#&gt; + 6/794 vertices, named, from 775cff6:</span>
<span class="co">#&gt; [1] ADAMTS9-AS2 C8orf34-AS1 CADM3-AS1   FAM83A-AS1  FENDRR      LANCL1-AS1</span></code></pre></div>
</div>
<div id="degree-centrality" class="section level2">
<h2 class="hasAnchor">
<a href="#degree-centrality" class="anchor" aria-hidden="true"></a>Degree centrality</h2>
<p>Degree centrality is the most commonly used local centrality measure which could be calculated via the <code>degree</code> function obtained from the <code>igraph</code> package.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculating degree centrality</span>
<span class="va">My_graph_degree</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/degree.html">degree</a></span><span class="op">(</span><span class="va">My_graph</span>, v <span class="op">=</span> <span class="va">GraphVertices</span>, normalized <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> 

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">My_graph_degree</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;         172         121         168          26         189         176</span></code></pre></div>
<p>Degree centrality could be also calculated for <em>directed</em> graphs via specifying the <code>mode</code> parameter.</p>
</div>
<div id="betweenness-centrality" class="section level2">
<h2 class="hasAnchor">
<a href="#betweenness-centrality" class="anchor" aria-hidden="true"></a>Betweenness centrality</h2>
<p>Betweenness centrality, like degree centrality, is one of the most commonly used centrality measures but is representative of the global centrality of a node. This centrality metric could also be calculated using a function obtained from the <code>igraph</code> package.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculating betweenness centrality</span>
<span class="va">My_graph_betweenness</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/betweenness.html">betweenness</a></span><span class="op">(</span><span class="va">My_graph</span>, v <span class="op">=</span> <span class="va">GraphVertices</span>,    
                                    directed <span class="op">=</span> <span class="cn">FALSE</span>, normalized <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">My_graph_betweenness</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;   21719.857   28185.199   26946.625    2940.467   33333.369   21830.511</span></code></pre></div>
<p>Betweenness centrality could be also calculated for <em>directed</em> and/or <em>weighted</em> graphs via specifying the <code>directed</code> and <code>weights</code> parameters, respectively.</p>
</div>
<div id="neighborhood-connectivity" class="section level2">
<h2 class="hasAnchor">
<a href="#neighborhood-connectivity" class="anchor" aria-hidden="true"></a>Neighborhood connectivity</h2>
<p>Neighborhood connectivity is one of the other important centrality measures that reflect the semi-local centrality of a node. This centrality measure was first represented in a Science paper<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> in 2002 and is for the first time calculable in R environment via the <code>influential</code> package.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculating neighborhood connectivity</span>
<span class="va">neighrhood.co</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/neighborhood.connectivity.html">neighborhood.connectivity</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,    
                                           vertices <span class="op">=</span> <span class="va">GraphVertices</span>,
                                           mode <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">neighrhood.co</span><span class="op">)</span>
<span class="co">#&gt;  ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;   11.290698    4.983471    7.970238    3.000000   15.153439   13.465909</span></code></pre></div>
<p>Neighborhood connectivity could be also calculated for <em>directed</em> graphs via specifying the <code>mode</code> parameter.</p>
</div>
<div id="h-index" class="section level2">
<h2 class="hasAnchor">
<a href="#h-index" class="anchor" aria-hidden="true"></a>H-index</h2>
<p>H-index is H-index is another semi-local centrality measure that was inspired from its application in assessing the impact of researchers and is for the first time calculable in R environment via the <code>influential</code> package.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculating H-index</span>
<span class="va">h.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/h_index.html">h_index</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,    
                   vertices <span class="op">=</span> <span class="va">GraphVertices</span>,
                   mode <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">h.index</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;          11           9          11           2          12          12</span></code></pre></div>
<p>H-index could be also calculated for <em>directed</em> graphs via specifying the <code>mode</code> parameter.</p>
</div>
<div id="local-h-index" class="section level2">
<h2 class="hasAnchor">
<a href="#local-h-index" class="anchor" aria-hidden="true"></a>Local H-index</h2>
<p>Local H-index (LH-index) is a semi-local centrality measure and an improved version of H-index centrality that leverages the H-index to the second order neighbors of a node and is for the first time calculable in R environment via the <code>influential</code> package.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculating Local H-index</span>
<span class="va">lh.index</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/lh_index.html">lh_index</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,    
                   vertices <span class="op">=</span> <span class="va">GraphVertices</span>,
                   mode <span class="op">=</span> <span class="st">"all"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lh.index</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;        1165         446         994          34        1289        1265</span></code></pre></div>
<p>Local H-index could be also calculated for <em>directed</em> graphs via specifying the <code>mode</code> parameter.</p>
</div>
<div id="collective-influence" class="section level2">
<h2 class="hasAnchor">
<a href="#collective-influence" class="anchor" aria-hidden="true"></a>Collective Influence</h2>
<p>Collective Influence (CI) is a global centrality measure that calculates the product of the reduced degree (degree - 1) of a node and the total reduced degree of all nodes at a distance d from the node. This centrality measure is for the first time provided in an R package.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculating Collective Influence</span>
<span class="va">ci</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/collective.influence.html">collective.influence</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,    
                          vertices <span class="op">=</span> <span class="va">GraphVertices</span>,
                          mode <span class="op">=</span> <span class="st">"all"</span>, d<span class="op">=</span><span class="fl">3</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ci</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;        9918       70560       39078         675       10716        7350</span></code></pre></div>
<p>Collective Influence could be also calculated for <em>directed</em> graphs via specifying the <code>mode</code> parameter.</p>
</div>
<div id="clusterrank" class="section level2">
<h2 class="hasAnchor">
<a href="#clusterrank" class="anchor" aria-hidden="true"></a>ClusterRank</h2>
<p>ClusterRank is a local centrality measure that makes a connection between local and semi-local characteristics of a node and at the same time removes the negative effects of local clustering.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculating ClusterRank</span>
<span class="va">cr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clusterRank.html">clusterRank</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,    
                  vids <span class="op">=</span> <span class="va">GraphVertices</span>,
                  directed <span class="op">=</span> <span class="cn">FALSE</span>, loops <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cr</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;   63.459812    5.185675   21.111776    1.280000  135.098278   81.255195</span></code></pre></div>
<p>ClusterRank could be also calculated for <em>directed</em> graphs via specifying the <code>directed</code> parameter.</p>
<p><a href="#top">Back to top</a></p>
</div>
</div>
<div id="assessment-of-the-association-of-centrality-measures" class="section level1">
<h1 class="hasAnchor">
<a href="#assessment-of-the-association-of-centrality-measures" class="anchor" aria-hidden="true"></a>Assessment of the association of centrality measures</h1>
<div id="conditional-probability-of-deviation-from-means" class="section level2">
<h2 class="hasAnchor">
<a href="#conditional-probability-of-deviation-from-means" class="anchor" aria-hidden="true"></a>Conditional probability of deviation from means</h2>
<p>The function <code>cond.prob.analysis</code> assesses the conditional probability of deviation of two centrality measures (or any other two continuous variables) from their corresponding means in opposite directions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">centrality.measures</span>        

<span class="co"># Assessing the conditional probability</span>
<span class="va">My.conditional.prob</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cond.prob.analysis.html">cond.prob.analysis</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">MyData</span>,       
                                          nodes.colname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>,
                                          Desired.colname <span class="op">=</span> <span class="st">"BC"</span>,
                                          Condition.colname <span class="op">=</span> <span class="st">"NC"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">My.conditional.prob</span><span class="op">)</span>
<span class="co">#&gt; $ConditionalProbability</span>
<span class="co">#&gt; [1] 51.61871</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConditionalProbability_split.half.sample</span>
<span class="co">#&gt; [1] 51.73611</span></code></pre></div>
<ul>
<li>As you can see in the results, the whole data is also randomly splitted into half in order to further test the validity of conditional probability assessments.</li>
<li>
<em>The higher the conditional probability the more two centrality measures behave in contrary manners</em>.</li>
</ul>
</div>
<div id="nature-of-association-considering-dependent-and-independent" class="section level2">
<h2 class="hasAnchor">
<a href="#nature-of-association-considering-dependent-and-independent" class="anchor" aria-hidden="true"></a>Nature of association (considering dependent and independent)</h2>
<p>The function <code>double.cent.assess</code> could be used to automatically assess both the distribution mode of centrality measures (two continuous variables) and the nature of their association. The analyses done through this formula are as follows:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Normality assessment</strong>:
<ul>
<li>Variables with <strong>lower than</strong> 5000 observations: <em>Shapiro-Wilk test</em>
</li>
<li>Variables with <strong>over</strong> 5000 observations: <em>Anderson-Darling test</em> <br><br>
</li>
</ul>
</li>
<li>
<strong>Assessment of non-linear/non-monotonic correlation</strong>:
<ul>
<li>
<em>Non-linearity assessment</em>: Fitting a generalized additive model (GAM) with integrated smoothness approximations using the <code>mgcv</code> package <br><br>
</li>
<li>
<em>Non-monotonicity assessment</em>: Comparing the squared coefficients of the correlation based on Spearman’s rank correlation analysis and ranked regression test with non-linear splines.
<ul>
<li>Squared coefficient of Spearman’s rank correlation <strong>&gt;</strong> R-squared ranked regression with non-linear splines: <em>Monotonic</em>
</li>
<li>Squared coefficient of Spearman’s rank correlation <strong>&lt;</strong> R-squared ranked regression with non-linear splines: <em>Non-monotonic</em> <br><br>
</li>
</ul>
</li>
</ul>
</li>
<li>
<strong>Dependence assessment</strong>:
<ul>
<li>
<em>Hoeffding’s independence test</em>: Hoeffding’s test of independence is a test based on the population measure of deviation from independence which computes a D Statistics ranging from -0.5 to 1: Greater D values indicate a higher dependence between variables.</li>
<li>
<em>Descriptive non-linear non-parametric dependence test</em>: This assessment is based on non-linear non-parametric statistics (NNS) which outputs a dependence value ranging from 0 to 1. For further details please refer to the NNS R package<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>: Greater values indicate a higher dependence between variables. <br><br>
</li>
</ul>
</li>
<li>
<strong>Correlation assessment</strong>: As the correlation between most of the centrality measures follows a non-monotonic form, this part of the assessment is done based on the NNS statistics which itself calculates the correlation based on partial moments and outputs a correlation value ranging from -1 to 1. For further details please refer to the NNS R package. <br><br>
</li>
<li>
<strong>Assessment of conditional probability of deviation from means</strong> This step assesses the conditional probability of deviation of two centrality measures (or any other two continuous variables) from their corresponding means in opposite directions.
<ul>
<li>The independent centrality measure (variable) is considered as the condition variable and the other as the desired one.</li>
<li>As you will see in the results, the whole data is also randomly splitted into half in order to further test the validity of conditional probability assessments.</li>
<li>
<em>The higher the conditional probability the more two centrality measures behave in contrary manners</em>.</li>
</ul>
</li>
</ol>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">centrality.measures</span>        

<span class="co"># Association assessment</span>
<span class="va">My.metrics.assessment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/double.cent.assess.html">double.cent.assess</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">MyData</span>,       
                                            nodes.colname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>,
                                            dependent.colname <span class="op">=</span> <span class="st">"BC"</span>,
                                            independent.colname <span class="op">=</span> <span class="st">"NC"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">My.metrics.assessment</span><span class="op">)</span>
<span class="co">#&gt; $Summary_statistics</span>
<span class="co">#&gt;         BC NC</span>
<span class="co">#&gt; Min.              0.000000000                   1.2000</span>
<span class="co">#&gt; 1st Qu.           0.000000000                  66.0000</span>
<span class="co">#&gt; Median            0.000000000                 156.0000</span>
<span class="co">#&gt; Mean              0.005813357                 132.3443</span>
<span class="co">#&gt; 3rd Qu.           0.000340000                 179.3214</span>
<span class="co">#&gt; Max.              0.529464720                 192.0000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Normality_results</span>
<span class="co">#&gt;                               p.value</span>
<span class="co">#&gt; BC    1.415450e-50</span>
<span class="co">#&gt; NC 9.411737e-30</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Dependent_Normality</span>
<span class="co">#&gt; [1] "Non-normally distributed"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Independent_Normality</span>
<span class="co">#&gt; [1] "Non-normally distributed"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $GAM_nonlinear.nonmonotonic.results</span>
<span class="co">#&gt;      edf  p-value </span>
<span class="co">#&gt; 8.992406 0.000000 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Association_type</span>
<span class="co">#&gt; [1] "nonlinear-nonmonotonic"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $HoeffdingD_Statistic</span>
<span class="co">#&gt;         D_statistic P_value</span>
<span class="co">#&gt; Results  0.01770279   1e-08</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Dependence_Significance</span>
<span class="co">#&gt;                       Hoeffding</span>
<span class="co">#&gt; Results Significantly dependent</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $NNS_dep_results</span>
<span class="co">#&gt;         Correlation Dependence</span>
<span class="co">#&gt; Results  -0.7948106  0.8647164</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConditionalProbability</span>
<span class="co">#&gt; [1] 55.35386</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConditionalProbability_split.half.sample</span>
<span class="co">#&gt; [1] 55.90331</span></code></pre></div>
<p><strong>Note</strong>: It should also be noted that as a single regression line does not fit all models with a certain degree of freedom, based on the size and correlation mode of the variables provided, this function might return an error due to incapability of running step 2. In this case, you may follow each step manually or as an alternative run the other function named <code>double.cent.assess.noRegression</code> which does not perform any regression test and consequently it is not required to determine the dependent and independent variables.</p>
</div>
<div id="nature-of-association-without-considering-dependence-direction" class="section level2">
<h2 class="hasAnchor">
<a href="#nature-of-association-without-considering-dependence-direction" class="anchor" aria-hidden="true"></a>Nature of association (without considering dependence direction)</h2>
<p>The function <code>double.cent.assess.noRegression</code> could be used to automatically assess both the distribution mode of centrality measures (two continuous variables) and the nature of their association. The analyses done through this formula are as follows:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Normality assessment</strong>:
<ul>
<li>Variables with <strong>lower than</strong> 5000 observations: <em>Shapiro-Wilk test</em>
</li>
<li>Variables with <strong>over</strong> 5000 observations: <em>Anderson–Darling test</em> <br><br>
</li>
</ul>
</li>
<li>
<strong>Dependence assessment</strong>:
<ul>
<li>
<em>Hoeffding’s independence test</em>: Hoeffding’s test of independence is a test based on the population measure of deviation from independence which computes a D Statistics ranging from -0.5 to 1: Greater D values indicate a higher dependence between variables.</li>
<li>
<em>Descriptive non-linear non-parametric dependence test</em>: This assessment is based on non-linear non-parametric statistics (NNS) which outputs a dependence value ranging from 0 to 1. For further details please refer to the NNS R package: Greater values indicate a higher dependence between variables. <br><br>
</li>
</ul>
</li>
<li>
<strong>Correlation assessment</strong>: As the correlation between most of the centrality measures follows a non-monotonic form, this part of the assessment is done based on the NNS statistics which itself calculates the correlation based on partial moments and outputs a correlation value ranging from -1 to 1. For further details please refer to the NNS R package. <br><br>
</li>
<li>
<strong>Assessment of conditional probability of deviation from means</strong> This step assesses the conditional probability of deviation of two centrality measures (or any other two continuous variables) from their corresponding means in opposite directions.
<ul>
<li>The <code>centrality2</code> variable is considered as the condition variable and the other (<code>centrality1</code>) as the desired one.</li>
<li>As you will see in the results, the whole data is also randomly splitted into half in order to further test the validity of conditional probability assessments.</li>
<li>
<em>The higher the conditional probability the more two centrality measures behave in contrary manners</em>.</li>
</ul>
</li>
</ol>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">centrality.measures</span>        

<span class="co"># Association assessment</span>
<span class="va">My.metrics.assessment</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/double.cent.assess.noRegression.html">double.cent.assess.noRegression</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">MyData</span>,       
                                                         nodes.colname <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>,
                                                         centrality1.colname <span class="op">=</span> <span class="st">"BC"</span>,
                                                         centrality2.colname <span class="op">=</span> <span class="st">"NC"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">My.metrics.assessment</span><span class="op">)</span>
<span class="co">#&gt; $Summary_statistics</span>
<span class="co">#&gt;         BC NC</span>
<span class="co">#&gt; Min.              0.000000000                   1.2000</span>
<span class="co">#&gt; 1st Qu.           0.000000000                  66.0000</span>
<span class="co">#&gt; Median            0.000000000                 156.0000</span>
<span class="co">#&gt; Mean              0.005813357                 132.3443</span>
<span class="co">#&gt; 3rd Qu.           0.000340000                 179.3214</span>
<span class="co">#&gt; Max.              0.529464720                 192.0000</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Normality_results</span>
<span class="co">#&gt;                               p.value</span>
<span class="co">#&gt; BC    1.415450e-50</span>
<span class="co">#&gt; NC 9.411737e-30</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Centrality1_Normality</span>
<span class="co">#&gt; [1] "Non-normally distributed"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Centrality2_Normality</span>
<span class="co">#&gt; [1] "Non-normally distributed"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $HoeffdingD_Statistic</span>
<span class="co">#&gt;         D_statistic P_value</span>
<span class="co">#&gt; Results  0.01770279   1e-08</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Dependence_Significance</span>
<span class="co">#&gt;                       Hoeffding</span>
<span class="co">#&gt; Results Significantly dependent</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $NNS_dep_results</span>
<span class="co">#&gt;         Correlation Dependence</span>
<span class="co">#&gt; Results  -0.7948106  0.8647164</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConditionalProbability</span>
<span class="co">#&gt; [1] 55.35386</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $ConditionalProbability_split.half.sample</span>
<span class="co">#&gt; [1] 55.68163</span></code></pre></div>
<p><a href="#top">Back to top</a></p>
</div>
</div>
<div id="IVI" class="section level1">
<h1 class="hasAnchor">
<a href="#IVI" class="anchor" aria-hidden="true"></a>Identification of the most <code>influential</code> network nodes</h1>
<p><strong>IVI</strong> : <code>IVI</code> is the first integrative method for the identification of network most influential nodes in a way that captures all network topological dimensions. The <code>IVI</code> formula integrates the most important local (i.e. degree centrality and ClusterRank), semi-local (i.e. neighborhood connectivity and local H-index) and global (i.e. betweenness centrality and collective influence) centrality measures in such a way that both synergizes their effects and removes their biases.</p>
<div id="integrated-value-of-influence-ivi-from-centrality-measures" class="section level2">
<h2 class="hasAnchor">
<a href="#integrated-value-of-influence-ivi-from-centrality-measures" class="anchor" aria-hidden="true"></a>Integrated Value of Influence (IVI) from centrality measures</h2>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">centrality.measures</span>        

<span class="co"># Calculation of IVI</span>
<span class="va">My.vertices.IVI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ivi.from.indices.html">ivi.from.indices</a></span><span class="op">(</span>DC <span class="op">=</span> <span class="va">MyData</span><span class="op">$</span><span class="va">DC</span>,       
                                   CR <span class="op">=</span> <span class="va">MyData</span><span class="op">$</span><span class="va">CR</span>,
                                   NC <span class="op">=</span> <span class="va">MyData</span><span class="op">$</span><span class="va">NC</span>,
                                   LH_index <span class="op">=</span> <span class="va">MyData</span><span class="op">$</span><span class="va">LH_index</span>,
                                   BC <span class="op">=</span> <span class="va">MyData</span><span class="op">$</span><span class="va">BC</span>,
                                   CI <span class="op">=</span> <span class="va">MyData</span><span class="op">$</span><span class="va">CI</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">My.vertices.IVI</span><span class="op">)</span>
<span class="co">#&gt; [1] 24.670056  8.344337 18.621049  1.017768 29.437028 33.512598</span></code></pre></div>
</div>
<div id="integrated-value-of-influence-ivi-from-a-graph" class="section level2">
<h2 class="hasAnchor">
<a href="#integrated-value-of-influence-ivi-from-a-graph" class="anchor" aria-hidden="true"></a>Integrated Value of Influence (IVI) from a graph</h2>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculation of IVI</span>
<span class="va">My.vertices.IVI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ivi.html">ivi</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>, vertices <span class="op">=</span> <span class="va">GraphVertices</span>, 
                       weights <span class="op">=</span> <span class="cn">NULL</span>, directed <span class="op">=</span> <span class="cn">FALSE</span>, mode <span class="op">=</span> <span class="st">"all"</span>,
                       loops <span class="op">=</span> <span class="cn">TRUE</span>, d <span class="op">=</span> <span class="fl">3</span>, scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">My.vertices.IVI</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;    39.53878    19.94999    38.20524     1.12371   100.00000    47.49356</span></code></pre></div>
<p>IVI could be also calculated for <em>directed</em> and/or <em>weighted</em> graphs via specifying the <code>directed</code>, <code>mode</code>, and <code>weights</code> parameters.</p>
<p>Check out <a href="https://doi.org/10.1016/j.patter.2020.100052" class="external-link">our paper</a><a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a> for a more complete description of the IVI formula and all of its underpinning methods and analyses.</p>
<p>The following tutorial video demonstrates how to simply calculate the IVI value of all of the nodes within a network.</p>
<iframe width="100%" height="400" src="https://www.youtube.com/embed/nF9qe4dvuJc%0A" frameborder="0" allowfullscreen>
</iframe>
</div>
<div id="NetVis" class="section level2">
<h2 class="hasAnchor">
<a href="#NetVis" class="anchor" aria-hidden="true"></a>Network visualization</h2>
<p>The <code>cent_network.vis</code> is a function for the visualization of a network based on applying a centrality measure to the size and color of network nodes. The centrality of network nodes could be calculated by any means and based on any centrality index. Here, we demonstrate the visualization of a network according to <a href="#IVI">IVI</a> values.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Reconstructing the graph</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">70</span><span class="op">)</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span>  <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/sample_gnm.html" class="external-link">sample_gnm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">50</span>, m <span class="op">=</span> <span class="fl">120</span>, directed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Calculating the IVI values</span>
<span class="va">My_graph_IVI</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ivi.html">ivi</a></span><span class="op">(</span><span class="va">My_graph</span>, directed <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># Visualizing the graph based on IVI values</span>
<span class="va">My_graph_IVI_Vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cent_network.vis.html">cent_network.vis</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,
                                     cent.metric <span class="op">=</span> <span class="va">My_graph_IVI</span>,
                                     directed <span class="op">=</span> <span class="cn">TRUE</span>,
                                     plot.title <span class="op">=</span> <span class="st">"IVI-based Network"</span>,
                                     legend.title <span class="op">=</span> <span class="st">"IVI value"</span><span class="op">)</span>

<span class="va">My_graph_IVI_Vis</span></code></pre></div>
<p><img src="../man/figures/IVI_Net_Vis.png"></p>
<p>The above figure illustrates a simple use case of the function <code>cent_network.vis</code>. You can apply this function to directed/undirected and/or weighted/unweighted networks. Also, a complete flexibility (list of arguments) have been provided for the adjustment of colors, transparencies, sizes, titles, etc. Additionally, several different layouts have been provided that could be conveniently applied to a network.</p>
<blockquote>
<p>In the case of highly crowded networks, the “<strong>grid</strong>” layout would be most appropriate.</p>
</blockquote>
<p>The following tutorial video demonstrates how to visualize a network based on the centrality of nodes (e.g. their <code>IVI</code> values).</p>
<iframe width="100%" height="400" src="https://www.youtube.com/embed/kCmEdOoIUD4" frameborder="0" allowfullscreen>
</iframe>
</div>
<div id="IVIShiny" class="section level2">
<h2 class="hasAnchor">
<a href="#IVIShiny" class="anchor" aria-hidden="true"></a>IVI shiny app</h2>
<p>A shiny app has also been developed for the calculation of IVI as well as IVI-based network visualization, which is accessible using the following command.<br><code>influential::runShinyApp("IVI")</code><br>
You can also access the shiny app online at the <a href="https://influential.erc.monash.edu/" class="external-link">Influential Software Package server</a>.</p>
<p><a href="#top">Back to top</a></p>
</div>
</div>
<div id="SpreadingScore" class="section level1">
<h1 class="hasAnchor">
<a href="#SpreadingScore" class="anchor" aria-hidden="true"></a>Identification of the most important network spreaders</h1>
<p>Sometimes we seek to identify not necessarily the most influential nodes but the nodes with most potential in spreading of information throughout the network.</p>
<p><strong>Spreading score</strong> : <code>spreading.score</code> is an integrative score made up of four different centrality measures including ClusterRank, neighborhood connectivity, betweenness centrality, and collective influence. Also, Spreading score reflects the spreading potential of each node within a network and is one of the major components of the <a href="#IVI"><code>IVI</code></a>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculation of Spreading score</span>
<span class="va">Spreading.score</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spreading.score.html">spreading.score</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,     
                                   vertices <span class="op">=</span> <span class="va">GraphVertices</span>, 
                                   weights <span class="op">=</span> <span class="cn">NULL</span>, directed <span class="op">=</span> <span class="cn">FALSE</span>, mode <span class="op">=</span> <span class="st">"all"</span>,
                                   loops <span class="op">=</span> <span class="cn">TRUE</span>, d <span class="op">=</span> <span class="fl">3</span>, scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Spreading.score</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;   42.932497   38.094111   45.114648    1.587262  100.000000   49.193292 </span></code></pre></div>
<p>Spreading score could be also calculated for <em>directed</em> and/or <em>weighted</em> graphs via specifying the <code>directed</code>, <code>mode</code>, and <code>weights</code> parameters. The results could be conveniently illustrated using the <a href="#NetVis">centrality-based network visualization function</a>.</p>
<p><a href="#top">Back to top</a></p>
</div>
<div id="HubnessScore" class="section level1">
<h1 class="hasAnchor">
<a href="#HubnessScore" class="anchor" aria-hidden="true"></a>Identification of the most important network hubs</h1>
<p>In some cases we want to identify not the nodes with the most sovereignty in their surrounding local environments.</p>
<p><strong>Hubness score</strong> : <code>hubness.score</code> is an integrative score made up of two different centrality measures including degree centrality and local H-index. Also, Hubness score reflects the power of each node in its surrounding environment and is one of the major components of the <a href="#IVI"><code>IVI</code></a>.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Preparing the data</span>
<span class="va">MyData</span> <span class="op">&lt;-</span> <span class="va">coexpression.data</span>        

<span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/graph_from_data_frame.html">graph_from_data_frame</a></span><span class="op">(</span><span class="va">MyData</span><span class="op">)</span>        

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculation of Hubness score</span>
<span class="va">Hubness.score</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/hubness.score.html">hubness.score</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,     
                                   vertices <span class="op">=</span> <span class="va">GraphVertices</span>, 
                                   directed <span class="op">=</span> <span class="cn">FALSE</span>, mode <span class="op">=</span> <span class="st">"all"</span>,
                                   loops <span class="op">=</span> <span class="cn">TRUE</span>, scaled <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Hubness.score</span><span class="op">)</span>
<span class="co">#&gt; ADAMTS9-AS2 C8orf34-AS1   CADM3-AS1  FAM83A-AS1      FENDRR  LANCL1-AS1 </span>
<span class="co">#&gt;   84.299719   46.741660   77.441514    8.437142   92.870451   88.734131</span></code></pre></div>
<p>Spreading score could be also calculated for <em>directed</em> graphs via specifying the <code>directed</code> and <code>mode</code> parameters. The results could be conveniently illustrated using the <a href="#NetVis">centrality-based network visualization function</a>.</p>
<p><a href="#top">Back to top</a></p>
</div>
<div id="SIRIR" class="section level1">
<h1 class="hasAnchor">
<a href="#SIRIR" class="anchor" aria-hidden="true"></a>Ranking the influence of nodes on the topology of a network based on the <code>SIRIR</code> model</h1>
<p><strong>SIRIR</strong> : <code>SIRIR</code> is achieved by the integration of susceptible-infected-recovered (SIR) model with the leave-one-out cross validation technique and ranks network nodes based on their true universal influence on the network topology and spread of information. One of the applications of this function is the assessment of performance of a novel algorithm in identification of network influential nodes.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Reconstructing the graph</span>
<span class="va">My_graph</span> <span class="op">&lt;-</span>  <span class="fu"><a href="../reference/sif2igraph.html">sif2igraph</a></span><span class="op">(</span>Path <span class="op">=</span> <span class="st">"Sample_SIF.sif"</span><span class="op">)</span>       

<span class="co"># Extracting the vertices</span>
<span class="va">GraphVertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My_graph</span><span class="op">)</span>        

<span class="co"># Calculation of influence rank</span>
<span class="va">Influence.Ranks</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sirir.html">sirir</a></span><span class="op">(</span>graph <span class="op">=</span> <span class="va">My_graph</span>,     
                                   vertices <span class="op">=</span> <span class="va">GraphVertices</span>, 
                                   beta <span class="op">=</span> <span class="fl">0.5</span>, gamma <span class="op">=</span> <span class="fl">1</span>, no.sim <span class="op">=</span> <span class="fl">10</span>, seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">difference.value</th>
<th align="right">rank</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">MRAP</td>
<td align="right">49.7</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td align="left">FOXM1</td>
<td align="right">49.5</td>
<td align="right">2</td>
</tr>
<tr class="odd">
<td align="left">ATAD2</td>
<td align="right">49.5</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">POSTN</td>
<td align="right">49.4</td>
<td align="right">4</td>
</tr>
<tr class="odd">
<td align="left">CDC7</td>
<td align="right">49.3</td>
<td align="right">5</td>
</tr>
<tr class="even">
<td align="left">ZWINT</td>
<td align="right">42.1</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">MKI67</td>
<td align="right">41.9</td>
<td align="right">7</td>
</tr>
<tr class="even">
<td align="left">FN1</td>
<td align="right">41.9</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">ASPM</td>
<td align="right">41.8</td>
<td align="right">9</td>
</tr>
<tr class="even">
<td align="left">ANLN</td>
<td align="right">41.8</td>
<td align="right">9</td>
</tr>
</tbody>
</table>
<p><a href="#top">Back to top</a></p>
</div>
<div id="ExIR" class="section level1">
<h1 class="hasAnchor">
<a href="#ExIR" class="anchor" aria-hidden="true"></a>Experimental data-based classification and ranking of top candidate features</h1>
<p><strong>ExIR</strong> : <code>ExIR</code> is a model for the classification and ranking of top candidate features. The input data could come from any type of experiment such as transcriptomics and proteomics. This model is based on multi-level filtration and scoring based on several supervised and unsupervised analyses followed by the classification and integrative ranking of top candidate features. Using this function and depending on the input data and specified arguments, the user can get a graph object and one to four tables including:</p>
<ul>
<li>
<strong>Drivers</strong>: Prioritized drivers are supposed to have the highest impact on the progression of a biological process or disease under investigation.</li>
<li>
<strong>Biomarkers</strong>: Prioritized biomarkers are supposed to have the highest sensitivity to different conditions under investigation and the severity of each condition.</li>
<li>
<strong>DE-mediators</strong>: Prioritized DE-mediators are those features that are differentially expressed/abundant but in a fluctuating manner and play mediatory roles between drivers.</li>
<li>
<strong>nonDE-mediators</strong>: Prioritized nonDE-mediators are those features that are not differentially expressed/abundant but have associations with and play mediatory roles between drivers.</li>
</ul>
<p>First, prepare your data. Suppose we have the data for time-course transcriptomics and we have previously performed differential expression analysis for each step-wise pair of time-points. Also, we have performed trajectory analysis to identify the genes that have significant alterations across all time-points.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Prepare sample data</span>
<span class="va">gene.names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"gene"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20000</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">60</span><span class="op">)</span>
<span class="va">tp2.vs.tp1.DEGs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>logFC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">700</span>, mean <span class="op">=</span> <span class="fl">2</span>, sd <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>,
                              FDR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">700</span>, min <span class="op">=</span> <span class="fl">0.0001</span>, max <span class="op">=</span> <span class="fl">0.049</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">60</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tp2.vs.tp1.DEGs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">gene.names</span>, size <span class="op">=</span> <span class="fl">700</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">70</span><span class="op">)</span>
<span class="va">tp3.vs.tp2.DEGs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>logFC <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1300</span>, mean <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, sd <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>,
                              FDR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1300</span>, min <span class="op">=</span> <span class="fl">0.0011</span>, max <span class="op">=</span> <span class="fl">0.039</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">70</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tp3.vs.tp2.DEGs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">gene.names</span>, size <span class="op">=</span> <span class="fl">1300</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">80</span><span class="op">)</span>
<span class="va">regression.data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>R_squared <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">800</span>, min <span class="op">=</span> <span class="fl">0.1</span>, max <span class="op">=</span> <span class="fl">0.85</span><span class="op">)</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">80</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">regression.data</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">gene.names</span>, size <span class="op">=</span> <span class="fl">800</span><span class="op">)</span></code></pre></div>
<div id="assembling-the-diff_data" class="section level2">
<h2 class="hasAnchor">
<a href="#assembling-the-diff_data" class="anchor" aria-hidden="true"></a>Assembling the Diff_data</h2>
<p>Use the function <code>diff_data.assembly</code> to automatically generate the Diff_data table for the <code>ExIR</code> model.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">my_Diff_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/diff_data.assembly.html">diff_data.assembly</a></span><span class="op">(</span><span class="va">tp2.vs.tp1.DEGs</span>,
                                   <span class="va">tp3.vs.tp2.DEGs</span>,
                                   <span class="va">regression.data</span><span class="op">)</span>

<span class="va">my_Diff_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span>,<span class="op">]</span></code></pre></div>
<p>Have a look at the top 10 rows of the <code>Diff_data</code> data frame:</p>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Diff_value1</th>
<th align="right">Sig_value1</th>
<th align="right">Diff_value2</th>
<th align="right">Sig_value2</th>
<th align="right">Diff_value3</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">gene_17331</td>
<td align="right">4.9</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">gene_12546</td>
<td align="right">4.0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">gene_12837</td>
<td align="right">-0.3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">gene_18522</td>
<td align="right">1.4</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">gene_6260</td>
<td align="right">-4.9</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">gene_2722</td>
<td align="right">-4.9</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">gene_19882</td>
<td align="right">6.3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">gene_2790</td>
<td align="right">3.3</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">gene_17011</td>
<td align="right">-1.6</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">gene_8321</td>
<td align="right">3.8</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
</div>
<div id="preparing-the-exptl_data" class="section level2">
<h2 class="hasAnchor">
<a href="#preparing-the-exptl_data" class="anchor" aria-hidden="true"></a>Preparing the Exptl_data</h2>
<p>Now, prepare a sample normalized experimental data matrix</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">60</span><span class="op">)</span>
<span class="va">MyExptl_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1000000</span>, min <span class="op">=</span> <span class="fl">2</span>, max <span class="op">=</span> <span class="fl">300</span><span class="op">)</span>, 
                       nrow <span class="op">=</span> <span class="fl">50</span>, ncol <span class="op">=</span> <span class="fl">20000</span>, 
                       dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"cancer_sample"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span>,
                                         <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"normal_sample"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">25</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span>, 
                                       <span class="va">gene.names</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Log transform the data to bring them closer to normal distribution</span>
<span class="va">MyExptl_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">MyExptl_data</span><span class="op">)</span>

<span class="va">MyExptl_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">45</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>Have a look at top 5 cancer and normal samples (rows) of the <code>Exptl_data</code>:</p>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">gene_1</th>
<th align="right">gene_2</th>
<th align="right">gene_3</th>
<th align="right">gene_4</th>
<th align="right">gene_5</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">cancer_sample_1</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">cancer_sample_2</td>
<td align="right">7</td>
<td align="right">8</td>
<td align="right">6</td>
<td align="right">8</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">cancer_sample_3</td>
<td align="right">8</td>
<td align="right">7</td>
<td align="right">8</td>
<td align="right">7</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">cancer_sample_4</td>
<td align="right">8</td>
<td align="right">7</td>
<td align="right">7</td>
<td align="right">7</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">cancer_sample_5</td>
<td align="right">6</td>
<td align="right">4</td>
<td align="right">8</td>
<td align="right">5</td>
<td align="right">4</td>
</tr>
<tr class="even">
<td align="left">normal_sample_20</td>
<td align="right">8</td>
<td align="right">7</td>
<td align="right">7</td>
<td align="right">8</td>
<td align="right">8</td>
</tr>
<tr class="odd">
<td align="left">normal_sample_21</td>
<td align="right">8</td>
<td align="right">7</td>
<td align="right">8</td>
<td align="right">6</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">normal_sample_22</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">7</td>
<td align="right">6</td>
</tr>
<tr class="odd">
<td align="left">normal_sample_23</td>
<td align="right">7</td>
<td align="right">6</td>
<td align="right">8</td>
<td align="right">7</td>
<td align="right">8</td>
</tr>
<tr class="even">
<td align="left">normal_sample_24</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">7</td>
<td align="right">5</td>
<td align="right">7</td>
</tr>
<tr class="odd">
<td align="left">normal_sample_25</td>
<td align="right">5</td>
<td align="right">7</td>
<td align="right">8</td>
<td align="right">8</td>
<td align="right">6</td>
</tr>
</tbody>
</table>
<p>Now add the “condition” column to the Exptl_data table.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">MyExptl_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">MyExptl_data</span><span class="op">)</span>
<span class="va">MyExptl_data</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"C"</span>, <span class="fl">25</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"N"</span>, <span class="fl">25</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="running-the-exir-model" class="section level2">
<h2 class="hasAnchor">
<a href="#running-the-exir-model" class="anchor" aria-hidden="true"></a>Running the <code>ExIR</code> model</h2>
<p>Finally, prepare the other required input data for the <code>ExIR</code> model.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co">#The table of differential/regression previously prepared</span>
<span class="va">my_Diff_data</span>

<span class="co">#The column indices of differential values in the Diff_data table</span>
<span class="va">Diff_value</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span> 

<span class="co">#The column indices of regression values in the Diff_data table</span>
<span class="va">Regr_value</span> <span class="op">&lt;-</span> <span class="fl">5</span>

<span class="co">#The column indices of significance (P-value/FDR) values in </span>
<span class="co"># the Diff_data table</span>
<span class="va">Sig_value</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span><span class="op">)</span> 

<span class="co">#The matrix/data frame of normalized experimental</span>
<span class="co"># data previously prepared</span>
<span class="va">MyExptl_data</span>

<span class="co">#The name of the column delineating the conditions of</span>
<span class="co"># samples in the Exptl_data matrix</span>
<span class="va">Condition_colname</span> <span class="op">&lt;-</span> <span class="st">"condition"</span>

<span class="co">#The desired list of features</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">60</span><span class="op">)</span>
<span class="va">MyDesired_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">gene.names</span>, size <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span>  <span class="co">#Optional</span>

<span class="co">#Running the ExIR model</span>
<span class="va">My.exir</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exir.html">exir</a></span><span class="op">(</span>Desired_list <span class="op">=</span> <span class="va">MyDesired_list</span>,
Diff_data <span class="op">=</span> <span class="va">my_Diff_data</span>, Diff_value <span class="op">=</span> <span class="va">Diff_value</span>,
Regr_value <span class="op">=</span> <span class="va">Regr_value</span>, Sig_value <span class="op">=</span> <span class="va">Sig_value</span>,
Exptl_data <span class="op">=</span> <span class="va">MyExptl_data</span>, Condition_colname <span class="op">=</span> <span class="va">Condition_colname</span>,
seed <span class="op">=</span> <span class="fl">60</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">My.exir</span><span class="op">)</span>
<span class="co">#&gt; [1] "Driver table"         "nonDE-mediator table" "Biomarker table"      "Graph"</span>

<span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">My.exir</span><span class="op">)</span>
<span class="co">#&gt; [1] "ExIR_Result"</span></code></pre></div>
<p>Have a look at the heads of the output tables of ExIR:</p>
<ul>
<li><strong>Drivers</strong></li>
</ul>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Score</th>
<th align="right">Z.score</th>
<th align="right">Rank</th>
<th align="right">P.value</th>
<th align="right">P.adj</th>
<th align="left">Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">gene_9469</td>
<td align="right">100</td>
<td align="right">13</td>
<td align="right">1</td>
<td align="right">1e-37</td>
<td align="right">0</td>
<td align="left">Accelerator</td>
</tr>
<tr class="even">
<td align="left">gene_339</td>
<td align="right">73</td>
<td align="right">9</td>
<td align="right">2</td>
<td align="right">2e-20</td>
<td align="right">0</td>
<td align="left">Accelerator</td>
</tr>
<tr class="odd">
<td align="left">gene_6518</td>
<td align="right">67</td>
<td align="right">8</td>
<td align="right">3</td>
<td align="right">1e-17</td>
<td align="right">0</td>
<td align="left">Decelerator</td>
</tr>
<tr class="even">
<td align="left">gene_13429</td>
<td align="right">60</td>
<td align="right">8</td>
<td align="right">4</td>
<td align="right">1e-14</td>
<td align="right">0</td>
<td align="left">Accelerator</td>
</tr>
<tr class="odd">
<td align="left">gene_8733</td>
<td align="right">58</td>
<td align="right">7</td>
<td align="right">5</td>
<td align="right">2e-13</td>
<td align="right">0</td>
<td align="left">Accelerator</td>
</tr>
<tr class="even">
<td align="left">gene_15640</td>
<td align="right">57</td>
<td align="right">7</td>
<td align="right">6</td>
<td align="right">3e-13</td>
<td align="right">0</td>
<td align="left">Accelerator</td>
</tr>
</tbody>
</table>
<hr>
<ul>
<li><strong>Biomarkers</strong></li>
</ul>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Score</th>
<th align="right">Z.score</th>
<th align="right">Rank</th>
<th align="right">P.value</th>
<th align="right">P.adj</th>
<th align="left">Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">gene_27</td>
<td align="right">100</td>
<td align="right">25.3</td>
<td align="right">1</td>
<td align="right">3e-141</td>
<td align="right">0.0</td>
<td align="left">Up-regulated</td>
</tr>
<tr class="even">
<td align="left">gene_2464</td>
<td align="right">32</td>
<td align="right">7.8</td>
<td align="right">2</td>
<td align="right">4e-15</td>
<td align="right">0.0</td>
<td align="left">Down-regulated</td>
</tr>
<tr class="odd">
<td align="left">gene_6903</td>
<td align="right">5</td>
<td align="right">0.9</td>
<td align="right">3</td>
<td align="right">2e-01</td>
<td align="right">0.5</td>
<td align="left">Up-regulated</td>
</tr>
<tr class="even">
<td align="left">gene_3196</td>
<td align="right">4</td>
<td align="right">0.7</td>
<td align="right">4</td>
<td align="right">2e-01</td>
<td align="right">0.5</td>
<td align="left">Up-regulated</td>
</tr>
<tr class="odd">
<td align="left">gene_13177</td>
<td align="right">3</td>
<td align="right">0.6</td>
<td align="right">5</td>
<td align="right">3e-01</td>
<td align="right">0.5</td>
<td align="left">Up-regulated</td>
</tr>
<tr class="even">
<td align="left">gene_2663</td>
<td align="right">3</td>
<td align="right">0.5</td>
<td align="right">6</td>
<td align="right">3e-01</td>
<td align="right">0.5</td>
<td align="left">Up-regulated</td>
</tr>
</tbody>
</table>
<hr>
<ul>
<li><strong>nonDE-mediators</strong></li>
</ul>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Score</th>
<th align="right">Z.score</th>
<th align="right">Rank</th>
<th align="right">P.value</th>
<th align="right">P.adj</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">gene_28</td>
<td align="right">100</td>
<td align="right">2.0</td>
<td align="right">1</td>
<td align="right">0.02</td>
<td align="right">0.1</td>
</tr>
<tr class="even">
<td align="left">gene_13648</td>
<td align="right">1</td>
<td align="right">-0.4</td>
<td align="right">2</td>
<td align="right">0.66</td>
<td align="right">0.7</td>
</tr>
<tr class="odd">
<td align="left">gene_19101</td>
<td align="right">1</td>
<td align="right">-0.4</td>
<td align="right">2</td>
<td align="right">0.66</td>
<td align="right">0.7</td>
</tr>
<tr class="even">
<td align="left">gene_15735</td>
<td align="right">1</td>
<td align="right">-0.4</td>
<td align="right">2</td>
<td align="right">0.66</td>
<td align="right">0.7</td>
</tr>
<tr class="odd">
<td align="left">gene_9637</td>
<td align="right">1</td>
<td align="right">-0.4</td>
<td align="right">2</td>
<td align="right">0.66</td>
<td align="right">0.7</td>
</tr>
<tr class="even">
<td align="left">gene_2841</td>
<td align="right">1</td>
<td align="right">-0.4</td>
<td align="right">2</td>
<td align="right">0.66</td>
<td align="right">0.7</td>
</tr>
</tbody>
</table>
<hr>
<ul>
<li><strong>nonDE-mediators</strong></li>
</ul>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">Score</th>
<th align="right">Z.score</th>
<th align="right">Rank</th>
<th align="right">P.value</th>
<th align="right">P.adj</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">gene_117</td>
<td align="right">100.00000</td>
<td align="right">5.236991</td>
<td align="right">1</td>
<td align="right">8.160804e-08</td>
<td align="right">0.0000071</td>
</tr>
<tr class="even">
<td align="left">gene_516</td>
<td align="right">57.59469</td>
<td align="right">2.693473</td>
<td align="right">2</td>
<td align="right">3.535593e-03</td>
<td align="right">0.1468437</td>
</tr>
<tr class="odd">
<td align="left">gene_262</td>
<td align="right">54.38275</td>
<td align="right">2.500817</td>
<td align="right">3</td>
<td align="right">6.195351e-03</td>
<td align="right">0.1468437</td>
</tr>
<tr class="even">
<td align="left">gene_974</td>
<td align="right">53.87269</td>
<td align="right">2.470223</td>
<td align="right">4</td>
<td align="right">6.751434e-03</td>
<td align="right">0.1468437</td>
</tr>
<tr class="odd">
<td align="left">gene_441</td>
<td align="right">46.60681</td>
<td align="right">2.034408</td>
<td align="right">5</td>
<td align="right">2.095524e-02</td>
<td align="right">0.3646212</td>
</tr>
<tr class="even">
<td align="left">gene_533</td>
<td align="right">40.91995</td>
<td align="right">1.693304</td>
<td align="right">6</td>
<td align="right">4.519882e-02</td>
<td align="right">0.6553829</td>
</tr>
</tbody>
</table>
<p>The following tutorial video demonstrates how to run the <code>ExIR</code> model on a sample experimental data.</p>
<iframe width="100%" height="400" src="https://www.youtube.com/embed/-08nn4_-OUE" frameborder="0" allowfullscreen>
</iframe>
<p>You can also <a href="#CompManipulate">computationally simulate knockout and/or up-regulation</a> of the top candidate features outputted by ExIR to evaluate the impact of their manipulations on the flow of information/signaling and integrity of the network prior to taking them to your lab bench.</p>
</div>
<div id="exir-visualization" class="section level2">
<h2 class="hasAnchor">
<a href="#exir-visualization" class="anchor" aria-hidden="true"></a>ExIR visualization</h2>
<p>The <code>exir.vis</code> is a function for the visualization of the output of the <a href="#ExIR">ExIR</a> model. The function simply gets the output of the ExIR model as a single argument and returns a plot of the top 10 prioritized features of all classes. Here, we visualize the top five candidates of the results of the <code>ExIR</code> model obtained in the previous step .</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">My.exir.Vis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/exir.vis.html">exir.vis</a></span><span class="op">(</span>exir.results <span class="op">=</span> <span class="va">My.exir</span>, 
                        n <span class="op">=</span> <span class="fl">5</span>,
                        y.axis.title <span class="op">=</span> <span class="st">"Gene"</span><span class="op">)</span>

<span class="va">My.exir.Vis</span></code></pre></div>
<p><img src="../man/figures/ExIR_Vis.png"></p>
<p>However, a complete flexibility (list of arguments) has been provided for the adjustment of all of the visual features of the plot and selection of the desired classes, feature types, and the number of top candidates.</p>
<p>The following tutorial video demonstrates how to visualize the results of <code>ExIR</code> model.</p>
<iframe width="100%" height="400" src="https://www.youtube.com/embed/Ie19y6PuBDg" frameborder="0" allowfullscreen>
</iframe>
</div>
<div id="ExIRShiny" class="section level2">
<h2 class="hasAnchor">
<a href="#ExIRShiny" class="anchor" aria-hidden="true"></a>ExIR shiny app</h2>
<p>A shiny app has also been developed for Running the ExIR model, visualization of its results as well as computational simulation of knockout and/or up-regulation of its top candidate outputs, which is accessible using the following command.<br><code>influential::runShinyApp("ExIR")</code><br>
You can also access the shiny app online at the <a href="https://influential.erc.monash.edu/" class="external-link">Influential Software Package server</a>.</p>
<p><a href="#top">Back to top</a></p>
</div>
</div>
<div id="CompManipulate" class="section level1">
<h1 class="hasAnchor">
<a href="#CompManipulate" class="anchor" aria-hidden="true"></a>Computational manipulation of cells</h1>
<p>The <code>comp_manipulate</code> is a function for the simulation of feature (gene, protein, etc.) knockout and/or up-regulation in cells. This function works based on the <a href="#SIRIR">SIRIR</a> (SIR-based Influence Ranking) model and could be applied on the output of the <a href="#ExIR">ExIR</a> model or any other independent association network. For feature (gene/protein/etc.) knockout the <a href="#SIRIR">SIRIR</a> model is used to remove the feature from the network and assess its impact on the flow of information (signaling) within the network. On the other hand, in case of up-regulation a node similar to the desired node is added to the network with exactly the same connections (edges) as of the original node. Next, the <a href="#SIRIR">SIRIR</a> model is used to evaluate the difference in the flow of information/signaling after adding (up-regulating) the desired feature/node compared with the original network. In case you are applying this function on the output of <a href="#ExIR">ExIR</a> model, you may note that as the gene/protein knockout would impact on the integrity of the under-investigation network as well as the networks of other overlapping biological processes/pathways, it is recommended to select those features that simultaneously have the highest (most significant) <a href="#ExIR">ExIR</a>-based rank and lowest knockout rank. In contrast, as the up-regulation would not affect the integrity of the network, you may select the features with highest (most significant) <a href="#ExIR">ExIR</a>-based and up-regulation-based ranks. Altogether, it is recommended to select the features with the highest (most significant) <a href="#ExIR">ExIR</a>-based (major drivers or mediators of the under-investigation biological process/disease) and <em>Up-regulation</em>-based (having higher impact on the signaling within the under-investigation network when up-regulated) ranks, but with the lowest <em>Knockout</em>-based rank (having the lowest disturbance to the under-investigation as well as other overlapping networks). Below is an example of running this function on the same <a href="#ExIR">ExIR</a> output generated above.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Select which genes to knockout</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">60</span><span class="op">)</span>
<span class="va">ko_vertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/as_ids.html" class="external-link">as_ids</a></span><span class="op">(</span><span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My.exir</span><span class="op">$</span><span class="va">Graph</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="co"># Select which genes to up-regulate</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span>
<span class="va">upregulate_vertices</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/as_ids.html" class="external-link">as_ids</a></span><span class="op">(</span><span class="fu"><a href="../reference/V.html">V</a></span><span class="op">(</span><span class="va">My.exir</span><span class="op">$</span><span class="va">Graph</span><span class="op">)</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>

<span class="va">Computational_manipulation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/comp_manipulate.html">comp_manipulate</a></span><span class="op">(</span>exir_output <span class="op">=</span> <span class="va">My.exir</span>, 
                                              ko_vertices <span class="op">=</span> <span class="va">ko_vertices</span>, 
                                              upregulate_vertices <span class="op">=</span> <span class="va">upregulate_vertices</span>,
                                              beta <span class="op">=</span> <span class="fl">0.5</span>, gamma <span class="op">=</span> <span class="fl">1</span>, no.sim <span class="op">=</span> <span class="fl">100</span>, seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></code></pre></div>
<p>Have a look at the heads of the output tables:</p>
<ul>
<li><strong>Knockout</strong></li>
</ul>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">Feature_name</th>
<th align="right">Rank</th>
<th align="left">Manipulation_type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">1</td>
<td align="left">gene_1877</td>
<td align="right">1</td>
<td align="left">Knockout</td>
</tr>
<tr class="even">
<td align="left">5</td>
<td align="left">gene_9469</td>
<td align="right">2</td>
<td align="left">Knockout</td>
</tr>
<tr class="odd">
<td align="left">4</td>
<td align="left">gene_2841</td>
<td align="right">3</td>
<td align="left">Knockout</td>
</tr>
<tr class="even">
<td align="left">2</td>
<td align="left">gene_13430</td>
<td align="right">4</td>
<td align="left">Knockout</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">gene_6197</td>
<td align="right">5</td>
<td align="left">Knockout</td>
</tr>
</tbody>
</table>
<ul>
<li><strong>Up-regulation</strong></li>
</ul>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">Feature_name</th>
<th align="right">Rank</th>
<th align="left">Manipulation_type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">5</td>
<td align="left">gene_2841</td>
<td align="right">1</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">gene_8733</td>
<td align="right">2</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="odd">
<td align="left">3</td>
<td align="left">gene_9469</td>
<td align="right">2</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="even">
<td align="left">2</td>
<td align="left">gene_13430</td>
<td align="right">4</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="odd">
<td align="left">4</td>
<td align="left">gene_16640</td>
<td align="right">5</td>
<td align="left">Up-regulation</td>
</tr>
</tbody>
</table>
<ul>
<li><strong>Combined</strong></li>
</ul>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="left">Feature_name</th>
<th align="right">Rank</th>
<th align="left">Manipulation_type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">51</td>
<td align="left">gene_2841</td>
<td align="right">1</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="even">
<td align="left">11</td>
<td align="left">gene_8733</td>
<td align="right">2</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="odd">
<td align="left">31</td>
<td align="left">gene_9469</td>
<td align="right">2</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="even">
<td align="left">21</td>
<td align="left">gene_13430</td>
<td align="right">4</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="odd">
<td align="left">41</td>
<td align="left">gene_16640</td>
<td align="right">5</td>
<td align="left">Up-regulation</td>
</tr>
<tr class="even">
<td align="left">1</td>
<td align="left">gene_1877</td>
<td align="right">6</td>
<td align="left">Knockout</td>
</tr>
</tbody>
</table>
<p><a href="#top">Back to top</a></p>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor" aria-hidden="true"></a>References</h1>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Csardi G., Nepusz T. <em>The igraph software package for complex network research</em>. InterJournal. 2006; (1695).<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Salavaty A, Rezvani Z, Najafi A. <em>Survival analysis and functional annotation of long non-coding RNAs in lung adenocarcinoma</em>. J Cell Mol Med. 2019;23:5600–5617. (<a href="https://pubmed.ncbi.nlm.nih.gov/31211495/" class="external-link">PMID: 31211495</a>)<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Maslov S., Sneppen K. <em>Specificity and stability in topology of protein networks</em>. Science. 2002; 296: 910-913 (<a href="https://pubmed.ncbi.nlm.nih.gov/11988575/" class="external-link">PMID:11988575</a>)<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>NNS: Nonlinear Nonparametric Statistics. (<a href="https://cran.r-project.org/package=NNS" class="external-link">CRAN</a>)<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>Salavaty A, Ramialison M, Currie PD. <em>Integrated Value of Influence: An Integrative Method for the Identification of the Most Influential Nodes within Networks</em>. Patterns. 2020.08.14. (<a href="https://doi.org/10.1016/j.patter.2020.100052" class="external-link">Read online</a>)<a href="#fnref5" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Adrian Salavaty.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
